File: 695-max-area-of-island/max-area-of-island.py
class Solution:
    def maxAreaOfIsland(self, grid: List[List[int]]) -> int:
        m,n = len(grid), len(grid[0])
        maxArea = 0 
        directions = [(1,0), (0,1), (-1,0), (0,-1)]
        
        visited = set()

        for i in range(m):
            for j in range(n):
                if grid[i][j] == 1 and (i,j) not in visited:
                    # do bfs here
                    count = 0
                    q = deque()
                    q.append((i,j))
                    visited.add((i,j))

                    while q:
                        x,y = q.pop()
                        count += 1
                        for dx,dy in directions:
                            newx, newy = x + dx, y + dy
                            if newx in range(m) and newy in range(n) and grid[newx][newy] == 1 and (newx, newy) not in visited:
                                q.append((newx, newy))
                                visited.add((newx, newy))

                    maxArea = max(maxArea, count)
        
        return maxArea

File: 695-max-area-of-island/max-area-of-island.cpp
class Solution {
public:
    int maxAreaOfIsland(vector<vector<int>>& grid) {
        int maxArea = 0;
        int m = grid.size(), n = grid[0].size();

        for (int i = 0; i < m; ++i) {
            for (int j = 0; j < n; ++j) {
                if (grid[i][j] == 1) {
                    maxArea = max(maxArea, bfs(grid, m, n, i, j));
                }
            }
        }
        return maxArea;
    }

private:
    int bfs(vector<vector<int>>& grid, int m, int n,  int r, int c) {
        vector<vector<int>> directions = {{0,1}, {1,0}, {0,-1}, {-1,0}};
        queue<pair<int,int>> q;

        q.push({r,c});
        grid[r][c] = 0;
        int area = 1;

        while (!q.empty()) {
            auto [row,col] = q.front();
            q.pop();

            for (const auto& dir : directions) {
                int x = dir[0];
                int y = dir[1];

                int nr = row + x;
                int nc = col + y;

                if (nr >= 0 && nr < m && nc >= 0 && nc < n && grid[nr][nc] == 1) {
                    grid[nr][nc] = 0;
                    q.push({nr,nc});
                    ++area;
                }
            }
        }
        return area;
    }
};